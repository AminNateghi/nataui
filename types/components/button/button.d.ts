import { ColorType, ShadowType, SizeType, VariantType } from '../../utilities/global-types';
import { PropsWithChildren } from 'react';
interface Props {
    size?: SizeType;
    variant?: VariantType;
    shadow?: ShadowType;
    color?: ColorType;
    disabled?: boolean;
    onClick?: () => void;
}
export declare const Button: ({ variant, size, shadow, color, disabled, onClick, children, }: Props & PropsWithChildren) => JSX.Element;
export {};
