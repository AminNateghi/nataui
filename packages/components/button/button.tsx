import React from 'react';
import { ColorType, ShadowType, SizeType, VariantType } from '../../utilities/global-types';
import { joinClassName } from '../../utilities/global-function';
import { PropsWithChildren } from 'react';

interface Props {
  size?: SizeType;
  variant?: VariantType;
  shadow?: ShadowType;
  color?: ColorType;
  disabled?: boolean;
  onClick?: () => void;
}

const Button = ({
  variant = 'outline',
  size = 'md',
  shadow,
  color,
  disabled = false,
  onClick,
  children,
}: Props & PropsWithChildren) => {
  const classNames = joinClassName(
    'button',
    `size-${size}`,
    shadow ? `shadow-${shadow}` : '',
    color ? `background-color-${color}` : `variant-${variant}`
  );
  return (
    <button
      className={classNames}
      disabled={disabled}
      onClick={onClick}
    >
      {children}
    </button>
  );
};

export default Button;
