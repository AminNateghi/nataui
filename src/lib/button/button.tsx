import { ColorType, ShadowType, SizeType, VariantType } from '../utils/global-types';
import { joinClassName } from '../utils/global-function';
import { PropsWithChildren } from 'react';

interface Props {
  size?: SizeType;
  variant?: VariantType;
  shadow?: ShadowType;
  color?: ColorType;
}

export const Button = ({
  variant = 'outline',
  size = 'md',
  shadow,
  color,
  children,
}: Props & PropsWithChildren) => {
  const classNames = joinClassName(
    'button',
    `size-${size}`,
    shadow ? `shadow-${shadow}` : '',
    color ? `background-color-${color}` : `variant-${variant}`,
    color ? `color-white` : ''
  );
  return <button className={classNames}>{children}</button>;
};
