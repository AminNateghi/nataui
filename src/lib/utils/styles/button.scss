.button {
  outline-color: transparent;
  // background-color: #f4f4f5;
  // color: #000;
  border-radius: 4px;
  transition: 0.3s;

  // button size
  // size-xs, size-sm, ...
  @include objectSize('xs', $xs-paddingX, $xs-fontSize, $xs-lineHeight, $xs-height, $xs-minWidth);
  @include objectSize('sm', $sm-paddingX, $sm-fontSize, $sm-lineHeight, $sm-height, $sm-minWidth);
  @include objectSize('md', $md-paddingX, $md-fontSize, $md-lineHeight, $md-height, $md-minWidth);
  @include objectSize('lg', $lg-paddingX, $lg-fontSize, $lg-lineHeight, $lg-height, $lg-minWidth);
  @include objectSize('xl', $xl-paddingX, $xl-fontSize, $xl-lineHeight, $xl-height, $xl-minWidth);

  // button variant
  // variant-outline, variant-simple, ...
  @include buttonVariant('outline', transparent, rgb(228, 228, 231), #000, false);
  @include buttonVariant('simple', #f4f4f5, transparent, #000, false);
  @include buttonVariant('solid', #f4f4f5, #d0d0d0, #000, false);
  @include buttonVariant('link', transparent, transparent, #000, true);

  // button box-shadow
  // shadow-sm, shadow-md, ...
  @include objectBoxShadow('xs');
  @include objectBoxShadow('sm');
  @include objectBoxShadow('md');
  @include objectBoxShadow('lg');
  @include objectBoxShadow('xl');
  @include objectBoxShadow('inner');
}