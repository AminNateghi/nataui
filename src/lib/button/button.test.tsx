import renderer from "react-test-renderer";
import { afterEach, describe, expect, test } from "vitest";
import { render, screen, cleanup } from "@testing-library/react";
import { Button } from "./button";

describe("Button - Unit test", () => {
  test("renders correctly", () => {
    const component = renderer.create(<Button />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("size prop", () => {
    const component = renderer.create(<Button size={"md"} />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("variant prop", () => {
    const component = renderer.create(<Button variant={"solid"} />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("Button - Integration test", () => {
  afterEach(cleanup);

  test("Expected label is displayed", () => {
    const label = "Hello, button";
    render(<Button>{label}</Button>);
    expect(screen.getByText(label));
  });
});
