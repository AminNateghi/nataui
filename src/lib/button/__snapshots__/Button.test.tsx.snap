// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button - Unit test > renders correctly 1`] = `
<button
  className="button size-md variant-outline"
/>
`;

exports[`Button - Unit test > size prop 1`] = `
<button
  className="button size-md variant-outline"
/>
`;

exports[`Button - Unit test > variant prop 1`] = `
<button
  className="button size-md variant-solid"
/>
`;
